source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -434,6 +445,22 @@ ReadOrParseState Filter::readProxyHeader(Network::IoHandle& io_handle) {
     if (nread < 1) {
       ENVOY_LOG(debug, "failed to read proxy protocol (no bytes read)");
       return ReadOrParseState::Error;
+    } else if (config_.get()->allowRequestsWithoutProxyProtocol() && buf_off_ == 0) {
+
+      if (nread < PROXY_PROTO_V1_SIGNATURE_LEN) {
```

Comment:
Sorry for I didn't describe the case clearly.

Write it as test case as below

```
TEST_P(ProxyProtocolTest, partialv1Signature) {

  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
  proto_config.set_allow_requests_without_proxy_protocol(true);
  connect(true, &proto_config);

  write("P");  // We only send single P here, it is less than `PROXY_PROTO_V1_SIGNATURE_LEN`, but the left of v1 signature will be send later. I think we shouldn't skip the left v1 signature parsing.

  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
  write("ROXY TCP4");
  write(" 254.254.2");
  write("54.254 1.2");
  write(".3.4 65535");
  write(" 1234\r\n...");

  expectData("...");

  EXPECT_EQ(server_connection_->connectionInfoProvider().remoteAddress()->ip()->addressAsString(),
            "254.254.254.254");
  EXPECT_TRUE(server_connection_->connectionInfoProvider().localAddressRestored());

  disconnect();
}
```

The test will fail with below:
```
test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc:160: Failure
Expected equality of these values:
  buffer.toString()
    Which is: "PROXY TCP4 254.254.254.254 1.2.3.4 65535 1234\r\n..."
  expected
    Which is: "..."
With diff:
@@ -1,2 @@
-PROXY TCP4 254.254.254.254 1.2.3.4 65535 1234\r
 ...
```

Reply:
see https://github.com/envoyproxy/envoy/pull/18951#discussion_r746888896

I'm not sure we can actually tell the difference between your case and the case provided above. If we receive a small amount of bytes off the initial `recv` we can be in either situation; I can change handling this scenario to a hard error, so that both of these cases see error rather than passing through a potentially valid proxy protocol request.

We could put `recv` in a loop with `MSG_PEEK` until we have enough bytes, but there's no guarantee there will ever be enough bytes and waiting for timeout is also bad for the [initial use case](https://github.com/solo-io/gloo/issues/5403#issuecomment-930529763).

Reply:
> I'm not sure we can actually tell the difference between your case and the case provided above. If we receive a small amount of bytes off the initial `recv` we can be in either situation; I can change handling this scenario to a hard error, so that both of these cases see error rather than passing through a potentially valid proxy protocol request.

What I'm thinking is if only see the first byte is a 'P', then we are looking for more bytes to ensure it is a proxy protocol. Not intent changing to a hard error.

> 
> We could put `recv` in a loop with `MSG_PEEK` until we have enough bytes, but there's no guarantee there will ever be enough bytes and waiting for timeout is also bad for the [initial use case](https://github.com/solo-io/gloo/issues/5403#issuecomment-930529763).

You are right, timeout is bad. It sounds like a choice. We can choice to believe after first byte 'P', there still can be a proxy protocol. Or we choice to believe after first byte 'P', it can't be a proxy protocol.

So I guess the question should be 'how much chance the healthcheck request is similar to proxy proxy signature?' If there is rarely chance the first few bytes of healthcheck is same with proxy signature, then I would choose to look more bytes to ensure it is proxy signature or not.

Reply:
I just realized that If we are going to fix the case I said, then we probably need to fix it with the way @ggreenway said at https://github.com/envoyproxy/envoy/pull/18951#discussion_r746945493

"I think you need to fix this case. You may have to restructure the code such that no bytes are read (without MSG_PEEK) until we've decided 100% whether this is a proxy header or not."

And I feel that will touch the windows MSG_PEEK issue, we have to wait for the fix  https://github.com/envoyproxy/envoy/pull/18900.

Reply:
Agreed, the PR as currently written does not read any bytes (without `MSG_PEEK`) until we know for sure whether or not this is proxy protocol.

Why do we have to wait for https://github.com/envoyproxy/envoy/pull/18900? We already use `MSG_PEEK` in this filter, and further the codepath added here is only hit if a user opts-in to the new feature. I'm only being slightly pushy because I have a customer chomping at the bits for this (eager user for feature).

Thanks and looking forward to your response.

Reply:
> Why do we have to wait for #18900? We already use `MSG_PEEK` in this filter, and further the codepath added here is only hit if a user opts-in to the new feature. I'm only being slightly pushy because I have a customer chomping at the bits for this (eager user for feature).
> 

The windows doesn't support edge trigger event, it emulated the edge trigger event by the level trigger event. It requires the caller always drain the data in the socket buffer, otherwise it always triggers the read event even there is no new data due to the level trigger event. Then it becomes a busy loop.

Reply:
Makes sense, but if I understand, this is already a problem in the proxy protocol filter as it stands today? My proposal here in this PR is to do the initial `MSG_PEEK` (as we have always done) and then act on the length of bytes received depending on our configuration.

We shouldn't be looping over `MSG_PEEK` without ever consuming bytes, and thus the behavior is largely unchanged. Further, the default behavior is literally unchanged.

Can we take another look at merging this?

Reply:
@mattklein123 can you weigh in here and provide direction? I believe we should try and make forward progress here instead of waiting on another PR dependency https://github.com/envoyproxy/envoy/pull/18900 which adds support for a required feature on Windows. I'm assuming we can add a feature for an OS and later on support it for other OS/distros/environments since this is opt in?

Reply:
Can someone summarize what needs deciding on please?

Reply:
Quick summary: this new feature would require doing MSG_PEEK to determine whether to proceed or not. The filter did not previously need to do this. On Windows, there isn't edge-triggered event support so this results in the connection having a read event on every event loop.

I don't think we can make a change that breaks Windows. It is a supported platform.

The fix for Windows is in the works. After that, this PR could proceed.

Reply:
>  I'm assuming we can add a feature for an OS and later on support it for other OS/distros/environments since this is opt in?

Yes, that is fine. The problem is that this change would break an existing feature on windows, until the other PR is merged.

Reply:
OK sounds good, let's just finish the other PR.